h/services/bulk_api/lms_stats.py

```diff
@@ -71,85 +77,102 @@ def _annotation_query(self, groups: list[str]):
+            .label("replies"),
+            func.max(counts_query.c.created).label("last_activity"),
+        )
 
-    def get_counts_by_assignment(self, groups: list[str]) -> list[CountsByAssignment]:
+    def get_annotation_counts(
+        self,
+        groups: list[str],
+        group_by: CountsGroupBy,
+        h_userids: list[str] | None = None,
+        assignment_id: str | None = None,
+    ) -> list[AnnotationCounts]:
         """
-        Get basic stats per assignment for an LMS course.
+        Get basic stats per user for an LMS assignment.
 
         :param groups: List of "authority_provided_id" to filter groups by.
+        :param group_by: By which column to aggregate the data.
+        :param h_userids: List of User.userid to filter annotations by
+        :param assignment_id: ID of the assignment to filter annotations by
         """
-        annos_query = self._annotation_query(groups).cte("annotations")
-        query = select(
-            annos_query.c.data["lms"]["assignment"]["resource_link_id"].astext.label(
-                "assignment_id"
+        annos_query = self._annotation_query(
+            groups, h_userids=h_userids, assignment_id=assignment_id
+        ).cte("annotations")
+
+        # Alias some columns
+        query_assignment_id = annos_query.c.data["lms"]["assignment"][
+            "resource_link_id"
+        ].astext
+        # Unfortunately all the magic around User.userid doesn't work in this context
+        query_userid = func.concat("acct:", User.username, "@", User.authority)
+
+        # What to group_by depending on the selection
+        group_by_clause = {
+            CountsGroupBy.USER: User.id,
+            CountsGroupBy.ASSIGNMENT: query_assignment_id,
+        }
+
+        # What columns to include, depending on the group by
+        group_by_select_columns = {
+            CountsGroupBy.USER: (
+                query_userid.label("userid"),
+                User.display_name,
             ),
-            func.count(annos_query.c.id)
-            .filter(annos_query.c.type == "annotation")
-            .label("annotations"),
-            func.count(annos_query.c.id)
-            .filter(annos_query.c.type == "reply")
-            .label("replies"),
-            func.max(annos_query.c.created).label("last_activity"),
-        ).group_by(annos_query.c.data["lms"]["assignment"]["resource_link_id"].astext)
+            CountsGroupBy.ASSIGNMENT: (query_assignment_id.label("assignment_id"),),
+        }
+
+        # What joins to include, depending on the group by
+        group_by_select_joins = {
+            CountsGroupBy.USER: ((annos_query, annos_query.c.user_id == User.id),),
+            CountsGroupBy.ASSIGNMENT: [],
+        }
+
+        query = select(
+            # Include the relevant columnns based on group_by
+            *group_by_select_columns[group_by],
+            # Always include the counts column
+            *self._count_columns(annos_query)
```

Comment:
Out of curiosity. What does the leading `*` do here?

Reply:
I passes annos_querires as positional arguments to the function. Two starts does the same but as named arguments, it can be used in dictionaries.